### PR TITLE
Feature/aksyonov/3 game logic

### PR DESCRIPTION
**Pull request by frozenstruct**

### CODE REVIEW

File: .github/workflows/ios.yml [hunk 1]
```diff
@@ -8,34 +8,40 @@ on:
 
 jobs:
   build:
-    name: Build and Test default scheme using any available iPhone simulator
+    name: xcodebuild + test on any iOS device
```
Comment (frozenstruct) [inline .github/workflows/ios.yml:11]:
```suggestion
    name: xcodebuild + test any iOS
```

File: .github/workflows/ios.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-name: iOS starter workflow
+name: xcodebuild + test
```
Comment (frozenstruct) [inline .github/workflows/ios.yml:1]:
```suggestion
name: Build For Tests
```

File: Numerica-iOS/Numerica/Sources/Storage/Objects/GameSessionObject.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//  Numerica
+//
+//  Created by Dmitry Aksyonov on 31.08.2024.
+//
+
+import Foundation
+import SwiftData
+
+@Model
+final class GameSessionObject {
```
Comment (frozenstruct) [inline Numerica-iOS/Numerica/Sources/Storage/Objects/GameSessionObject.swift:12]: Add game session type

File: Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+//  Created by Dmitry Aksyonov on 11.09.2024.
+//
+
+import Foundation
+import SwiftData
+
+@Model
+final class ProblemObject {
+
+    var id = UUID()
```
Comment (frozenstruct) [inline Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift:14]: See if we really need this

File: Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    var problem: String
+    var solution: Int
+    var solved: Bool
```
Comment (frozenstruct) [inline Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift:18]:
```suggestion
<<<<<<<<<<<<
    var problem: String
    var solution: Int
    var solved: Bool
>>>>>>>>>>>>

    let lhs: Int
    let operation: Operation
    let rhs: Int

    var problemString: String {
        [
         lhs,
         operation.rawValue,
         rhs
        ].joined(
          separator: .space
        )
    }

    init...

    // This will fix redundant string manipulation insure problem gen 
```

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            getRandomNumber(
+                shouldAvoidZero: operation == .division || divisionViaMultiplication
+            ).stringValue,
+            operation.rawValue,
+            getRandomNumber(
+                shouldAvoidZero: operation == .division || divisionViaMultiplication
+            ).stringValue
+        ].joined(separator: .space)
+
+        return ProblemDTO(
```
Comment (frozenstruct) [inline Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:105]: Consider comments on problem object and amend here Accordingly

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            if iteration < 25 {
+                operation = .addition
+            } else if iteration >= 25 {
+                operation = .subtraction
+            } else if iteration >= 50 {
+                operation = .multiplication
+            } else {
+                operation = .division
+            }
+
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:176]:
```suggestion

// Act
```

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            let generatedOperator = String(
+                sut.getProblem(
+                    for: operation
+                ).problemString.split(
+                    separator: .space
+                )[1]
+            )
+
+            XCTAssertEqual(operation.rawValue, generatedOperator)
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:185]:
```suggestion
// Assert
            XCTAssertEqual(operation.rawValue, generatedOperator)

```

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            ).problemString.split(
+                separator: .space
+            ).map(String.init)
+
+            guard let rhs = divisionProblem[2].intValue else {
+                XCTFail("Could not parse number from the given string")
+                return
+            }
+
+            XCTAssertTrue(rhs > .zero)
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:204]:
```suggestion
// Assert
            XCTAssertTrue(rhs > .zero)
```

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+            case 3:
+                XCTAssertEqual(value.rawValue, "/")
+            default:
+                XCTFail("Number of operations exceeds the needed number (4)")
+            }
+        }
+    }
+
+    func testOperationRandomness() {
+        let symbols = ["+", "-", "*", "/"]
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:231]:
```suggestion
// Arrange
        let symbols = ["+", "-", "*", "/"]
```

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        var numberOfOperations = 1000
+        var frequency = [String: Int]()
+        symbols.forEach {
+            frequency[$0] = 0
+        }
+
+        for _ in .zero...numberOfOperations {
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:238]:
```suggestion
// Act
        for _ in .zero...numberOfOperations {
```

File: Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+        symbols.forEach {
+            guard let observedFrequency = frequency[$0]?.doubleValue else {
+                XCTFail("Got operator that should not be generated")
+                return
+            }
+            chiSquare += pow(observedFrequency - expectedFrequency, 2) / expectedFrequency
+        }
+
+        let threshold = 7.81
+        print("chiSquare: \(chiSquare), threshold: \(threshold)")
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaTests/ProblemGeneratorTests.swift:267]:
```suggestion
// Assert
// FIXME: - make XCTASSERTEQUAL
        print("chiSquare: \(chiSquare), threshold: \(threshold)")
```

File: Numerica-iOS/NumericaUITests/NumericaTestsLaunchTests.swift [hunk 13]
```diff
@@ -0,0 +1,10 @@
+//
+//  NumericaUITestsLaunchTests.swift
+//  NumericaUITests
+//
+//  Created by Dmitry Aksyonov on 31.08.2024.
+//
+
+import XCTest
+
+final class NumericaUITestsLaunchTests: XCTestCase {}
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaUITests/NumericaTestsLaunchTests.swift:10]: Implement at least one test

File: Numerica-iOS/NumericaUITests/NumericaUITests.swift [hunk 14]
```diff
@@ -0,0 +1,10 @@
+//
+//  NumeriaUITests.swift
+//  NumeriaUITests
+//
+//  Created by Dmitry Aksyonov on 31.08.2024.
+//
+
+import XCTest
+
+final class NumeriaUITests: XCTestCase {}
```
Comment (frozenstruct) [inline Numerica-iOS/NumericaUITests/NumericaUITests.swift:10]: Try one too

**Review by frozenstruct — COMMENTED** [review]
Dumped a whit of comments

File: Docs/2_Problem Generation.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+3. Score tracking
+4. Statistics
+5. Analytics
+6. Insights
+
+will be stored in separate package written in [Rust](https://www.rust-lang.org). See technical consideration of selectig Rust as a primary language for game logic in [Core in Rust](./4_4_TDR%3A%20Core%20in%20Rust.md). Not only should the game logic be thoroughly encoded in a framework, it should have meaningful textual description to facilitate migration to Rust. Moreover, stated methodology mill allow for targetet unit-testing.
+
+## Enforcing Methodology in Code
+
+`XCTestCase` is a good option to enforce methodology thtrough the system of test-calses. For one, randomness ratios, edge cases ans conforming to methodology principles can be testet thoughout unit-tests. Having these tests in place will alert a developer each time he has broken the crucial logic.
```
Comment (frozenstruct) [inline Docs/2_Problem Generation.md:22]:
```suggestion
`XCTestCase` is a good option to enforce methodology through the system of test cases. For one, randomness ratios, edge cases and conforming to methodology principles can be tested via unit testing. Having these tests in place will alert a developer each time he has broken the crucial logic.
These tests should be structured according to the methodology described below. 
```suggestion

File: Docs/2_Problem Generation.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+## Problem Generation Methodology
+
+In general, problem generation flow can be enumerated by the follwing steps:
```
Comment (frozenstruct) [inline Docs/2_Problem Generation.md:26]:
```suggestion
Problem generation anatomy comprises of the following steps:
```

File: Docs/2_Problem Generation.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+1. Determining the operation
+2. Selecting digits capacity
+3. Defining the series' upper bound for random number selection
```
Comment (frozenstruct) [inline Docs/2_Problem Generation.md:30]:
```suggestion
3. Defining the upper and lower bounds for random number selection range
```

File: Docs/2_Problem Generation.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+4. Pikcing random number from the series
```
Comment (frozenstruct) [inline Docs/2_Problem Generation.md:31]:
```suggestion
4. Pikcing random number from the range
```

**Review by frozenstruct — COMMENTED** [review]
Dumped another shit load of comments

**Review by frozenstruct — COMMENTED** [review]
fixed some things

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+//  Created by Dmitry Aksyonov on 01.09.2024.
+//
+
+import Foundation
+
+// MARK: - MathGen
+
+final class ProblemGenerator {
+
+    private static var rng = SystemRandomNumberGenerator()
```
Comment (Wtclrsnd) [inline Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:14]: неочевидный нейминг по крайней мере для меня

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+// MARK: - Expression Solving
+
+private extension ProblemGenerator {
+
+    func solve(expression: String) -> Int {
+        let components = expression.split(
+            separator: .space
+        ).map(
```
Comment (Wtclrsnd) [inline Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:30]: ну чет тут не оч смотрятся переносы

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+            String.init
+        )
+
+        guard
+            let lhs = Int(components[.zero]),
```
Comment (Wtclrsnd) [inline Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:35]: а мы точно хотим завязываться на арр и по индексам тоскать все? может перепил на кортеж будет чище

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+            let rhs = String(components[2]).intValue
+        else {
+            return .empty
+        }
+
+        // Additional randomization of lhs / rhs
+        let randomNumber = Int.random(in: 0...1, using: &Self.rng)
+
+        let divisionProblemString = [
+            randomNumber == 0 ? rhs.stringValue : lhs.stringValue,
```
Comment (Wtclrsnd) [inline Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:80]: пиздец тяжелая операция на мой взгляд и тернарники можно убрать в мелкие функции или компьютед проперти. + я бы лучше сделал кортеж с именованными элементами

File: Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift (hunk 4 above)
Comment (Wtclrsnd) [reply Numerica-iOS/Numerica/Sources/Storage/Objects/ProblemObject.swift:14]: мне кажется лишним не будет

**Review by Wtclrsnd — COMMENTED** [review]
небольшие неочевидные моменты заметил, а так по всему коду в общем и тестам вопросов нет.

File: .github/workflows/ios.yml [hunk 23]
```diff
@@ -8,34 +8,40 @@ on:
...
 jobs:
   build:
-    name: Build and Test default scheme using any available iPhone simulator
+    name: xcodebuild + test any iOS
     runs-on: macos-latest
 
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+
```
Comment (frozenstruct) [inline .github/workflows/ios.yml:17]:
```suggestion
```

File: .github/workflows/ios.yml [hunk 24]
```diff
@@ -8,34 +8,40 @@ on:
...
       - name: Set Default Scheme
         run: |
+          cd Numerica-iOS
           scheme_list=$(xcodebuild -list -json | tr -d "\n")
           default=$(echo $scheme_list | ruby -e "require 'json'; puts JSON.parse(STDIN.gets)['project']['targets'][0]")
           echo $default | cat >default
           echo Using default scheme: $default
+
```
Comment (frozenstruct) [inline .github/workflows/ios.yml:25]:
```suggestion
```

File: .github/workflows/ios.yml [hunk 25]
```diff
@@ -8,34 +8,40 @@ on:
...
           platform: ${{ 'iOS Simulator' }}
         run: |
+          cd Numerica-iOS
           # xcrun xctrace returns via stderr, not the expected stdout (see https://developer.apple.com/forums/thread/663959)
           device=`xcrun xctrace list devices 2>&1 | grep -oE 'iPhone.*?[^\(]+' | head -1 | awk '{$1=$1;print}' | sed -e "s/ Simulator$//"`
           if [ $scheme = default ]; then scheme=$(cat default); fi
           if [ "`ls -A | grep -i \\.xcworkspace\$`" ]; then filetype_parameter="workspace" && file_to_build="`ls -A | grep -i \\.xcworkspace\$`"; else filetype_parameter="project" && file_to_build="`ls -A | grep -i \\.xcodeproj\$`"; fi
           file_to_build=`echo $file_to_build | awk '{$1=$1;print}'`
           xcodebuild build-for-testing -scheme "$scheme" -"$filetype_parameter" "$file_to_build" -destination "platform=$platform,name=$device"
+
```
Comment (frozenstruct) [inline .github/workflows/ios.yml:38]:
```suggestion
```

**Review by frozenstruct — COMMENTED** [review]
IMAGE 2024-
![IMAGE 2024-09-16 00:19:24
09-16 00:19:21](https://github.com/user-attachments/assets/1c26b5fe-dff0-461f-be42-88b0ddd32158)

IMAGE 2024-09-16 00:19:18

IMAGE 2024-09-16 00:19:14

IMAGE 2024-09-16 00:19:09

IMAGE 2024-09-16 00:19:02

**Review by frozenstruct — COMMENTED** [review]
reviewed

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift (hunk 22 above)
Comment (frozenstruct) [reply Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:80]: Исправлено в рамках исправлений выше

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift (hunk 20 above)
Comment (frozenstruct) [reply Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:30]: убрано в рамках исправлений выше

File: Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift (hunk 21 above)
Comment (frozenstruct) [reply Numerica-iOS/Numerica/Sources/Services/ProblemGenerator.swift:35]: убрано в рамках исправлений выше

**Review by frozenstruct — COMMENTED** [review]
addressed comments

**Review by frozenstruct — COMMENTED** [review]
pre-merge

**Review by Wtclrsnd — APPROVED** [review]
